torch/utils/checkpoint.py

```diff
@@ -335,7 +339,299 @@ def forward(input):
+            defaultdict(weakref.WeakKeyDictionary)
+        self.recomp_counter: DefaultDict[int, int] = defaultdict(int)
+        self.is_recomputed: DefaultDict[int, bool] = defaultdict(bool)
+
+# See Rule 5
+class _StopRecomputationError(Exception):
+    pass
+
+class _recomputation_hook(torch.autograd.graph.saved_tensors_hooks):
+    def __init__(self, target_frame_ref: ReferenceType, gid: int):
+        def pack_hook(x):
+            target_frame = target_frame_ref()
+            assert target_frame is not None
+            recomp_idx = target_frame.recomp_counter[gid]
+            target_frame.recomp_counter[gid] += 1
+
+            if recomp_idx >= len(target_frame.weak_holders):
+                # We run into this case when early stop is not enabled and do
+                # grad within checkpoint.
+                return x.detach()
+            holder = target_frame.weak_holders[recomp_idx]()
+
+            if holder is not None:
+                # See Rule 6: [ Multiple backwards ] above
+                if holder.handles.get(gid, None) is None:
+                    holder.handles[gid] = _Handle()
+                target_frame.recomputed[gid][holder.handles[gid]] = x.detach()
+
+            if _enable_checkpoint_early_stop and \
+               target_frame.recomp_counter[gid] == len(target_frame.weak_holders):
+                raise _StopRecomputationError()
+            # See Rule 6: [ Basic case ] above
+            return x.detach()
+
+        def unpack_hook(x):
+            return x
+
+        super().__init__(pack_hook, unpack_hook)
+
+class _checkpoint_hook(torch.autograd.graph.saved_tensors_hooks):
+    def __init__(self, frame):
+        def pack_hook(_unused_x):
+            # See Rule 4 above
+            holder = _Holder()
+            frame.weak_holders.append(weakref.ref(holder))
+            return holder
+
+        def unpack_hook(holder):
+            def do_checks(handle: _Handle, wkd: weakref.WeakKeyDictionary) -> None:
+                if handle is None:
+                    raise RuntimeError(
+                        "If you are calling ctx.saved_tensor in backward, make sure to do so only once. "
+                        "Otherwise please open an issue with details on your use case."
+                    )
+                if handle not in wkd:
+                    raise RuntimeError(
+                        "Attempt to retrieve a tensor saved by autograd multiple times without checkpoint"
+                        " recomputation being triggered in between, this is not currently supported. Please"
+                        " open an issue with details on your use case."
+                    )
+
+            gid = torch._C._current_graph_task_id()
+            if gid == -1:
+                # generate a temporary id if we trigger unpack outside of a backward call
+                gid = int(uuid.uuid4())
+
+            if frame.is_recomputed[gid]:
```

Comment:
Not that it is new in this impl, but we should add an issue to make this thread safe ;)

Reply:
I may be missing something, but doesn't the GIL protect this?

Reply:
Oh interesting, thanks for the clarification!